### ISSUE TITLE
Integration with koa middlewares.

### ISSUE THREAD

**Issue by aga5tya** [60frames/webpack-hot-server-middleware#28]
I'm trying to get library up with koa2 server based middleware, but i end up with the error below.

```

/Users/prithviraju/Sites/site/node_modules/memory-fs/lib/MemoryFileSystem.js:112
                        throw new MemoryFileSystemError(errors.code.EISDIR, _path);
         ^
Error
    at MemoryFileSystem.readFileSync (/Users/prithviraju/Sites/site/node_modules/memory-fs/lib/
MemoryFileSystem.js:112:10)
    at MultiCompiler.multiCompiler.plugin.multiStats (/Users/prithviraju/Sites/site/node_module
s/webpack-hot-server-middleware/src/index.js:119:33)
    at MultiCompiler.applyPlugins (/Users/prithviraju/Sites/site/node_modules/tapable/lib/Tapab
le.js:61:14)
```
I'm using koa connect library to achieve the compatibility by wrapping connect middleware, but stuck with the error. Its possibly related to 60frames/webpack-hot-server-middleware#7 and because of the function signature of entry point needed ```(stats) => (req, res, next) => void```, not sure how to make it a koa based one, tried wrapping in koa connect, but no avail.
Please point me in some direction.

**Comment by eudaimos** [comment]
First I want to say thanks to @richardscarrott for making this and providing it to the community.

I need the same thing, and I've forked the repo to try to provide a solution.

I have hacked the following solution into the file in `node_modules` in order to get my local setup working (for anyone who needs a quick solution now), which is backwards compatible, but not where I think this lib should go. (immediately I'll likely lift this to a place in my repo)

Replace [src/index.js:132-138](../blob/0.1.0/src/index.js#L132-L138) with the following:
```javascript
    const connectMiddleware = (req, res, next) => {
        debug(`Receive request ${req.url}`);
        if (error) {
            return next(error);
        }
        serverRenderer(req, res, next);
    };
    const koaMiddleware = (ctx, next) => {
        debug(`Receive request ${ctx.url}`);
        if (error) {
            ctx.throw(error);
        }
        serverRenderer(ctx, next);
    };
    connectMiddleware.koa = koaMiddleware;
    return connectMiddleware;
```

imho I don't believe this package needs to be tied to any middleware style b/c:
1. the only time the middleware signature is used when returning a middleware function to use that is wrapping the `serverRenderer` function that it dynamically looked up from the webpack build
2. it's really a good set of functions that enable this functionality that can be used in any server implementation

The possibilities for updating to allow `koa 2` middleware would be:
1. provide an alternate return (like above) but in a better configurable way
2. expose the various functions to allow anyone to create their own middleware function from the various parts
3. move to an Event/Promise/Observable model that would allow more composition in various ways with reference implementations

In the immediate, I don't have time to focus on this as I'm trying to get various pieces to a puzzle together, but very soon (next week) I will have to provide a package that can be used with the project on which I'm currently working, so I will be testing out my various _possibilities_ above

**Comment by richardscarrott** [comment]
@eudaimos thanks for taking a look into this -- I agree, this package doesn't really need to be tied to a particular middleware style as the only dependencies of the middleware function are the `error` and `serverRenderer` from the webpack compiler.

Based on your example we could accept the handler as an option, but ship with an express / connect and koa version as they're popular enough, e.g.

Replace [src/index.js:132-138](../blob/0.1.0/src/index.js#L132-L138) with the following:

```
return function() {
    return options.createHandler(error, serverRenderer)(...arguments);
};
```

And provide the following two handler creators (defaulting the options to createConnectHandler):

```
const createConnectHandler = (error, serverRenderer) => (req, res, next) => {
    debug(`Receive request ${req.url}`);
    if (error) {
        return next(error);
    }
    serverRenderer(req, res, next);
};

const createKoaHandler = (error, serverRenderer) => (ctx, next) => {
    debug(`Receive request ${ctx.url}`);
    if (error) {
        ctx.throw(error); // ASIDE: Does this need to return to ensure the `serverRenderer` isn't called when the middleware stack unwinds?
    }
    serverRenderer(ctx, next);
};
```

That way the consumer would do:

```
// Connect / Express
app.use(webpackHotServerMiddleware(compiler));

// Koa
app.use(webpackHotServerMiddleware(compiler, {
    createHandler: webpackHotServerMiddleware.createKoaHandler
}));

// Some other middleware library
app.use(webpackHotServerMiddleware(compiler, {
    createHandler: (error, serverRenderer) => (foo) => {
       if (error) {
           foo.handleError(error);
      }
      serverRenderer(foo);
   }
}));
```

**Comment by eudaimos** [comment]
awesome @richardscarrott thanks for supporting us!

I realized as I was debugging an annoying issue that I created myself (of course) but required me to go deep into the imported stack as I am integrating with a different library, that in haste my own `koa` handler code up there is wrong as I _excluded_ the `async` on the handler causing the koa handler to behave synchronously.

Updating your example code, it **should** look like this:
```javascript
const createKoaHandler = (error, serverRenderer) => async (ctx, next) => {
    debug(`Receive request ${ctx.url}`);
    if (error) {
        return ctx.throw(error); // ASIDE: Does this need to return to ensure the `serverRenderer` isn't called when the middleware stack unwinds? ANSWER: Probably
    }
    await serverRenderer(ctx, next);
};
```

Going to wire it up for myself locally the way you have it and run through some iterations.  Then I'll fork and create a PR.

**Comment by eudaimos** [comment]
As I'm looking at it @richardscarrott I don't think this will work b/c the execution context gets screwed up.  In the original implementation, and in the one I provided, the handlers are created inside of the body of `webpackHotServerMiddleware` giving them access to the live `error` value that gets updated on any webpack rebuild - we'll need to add some misdirection so that `error` and `serverRenderer` are functions that return the current version at the time the handler is executing.

**Comment by richardscarrott** [comment]
@eudaimos the `error` and `serverRenderer` should always be the latest values because the handler is wrapped in a function which re-creates the handler on every request, passing in the most up to date values, e.g.

```
return function() {
    return options.createHandler(error, serverRenderer)(...arguments);
};
```

**Comment by richardscarrott** [comment]
One problem I have noticed when hacking something together with koa is that `webpack-dev-middleware` is responding with a 404 when requesting static assets in the client -- even though the  requested file exists and is returned, e.g.

<img width="1225" alt="screen shot 2017-07-01 at 11 46 51" src="https://user-images.githubusercontent.com/539309/27761439-c04bd2e0-5e53-11e7-8433-dc17d715da9e.png">

**Comment by roastlechon** [comment]
@richardscarrott the issue you referenced is closed and PR merged https://github.com/webpack/webpack-dev-middleware/pull/175 on April 2017

Any progress @eudaimos?

**Comment by pho3nixf1re** [comment]
I need this support. If the proposed solution above works then I will submit a PR later this week.

**Comment by roastlechon** [comment]
I have something that works, but its not in a good place to merge.

Because I have babel-node running on the server, the class instanceof check of MultiCompiler is different than MultiCompiler from require('webpack/lib/MultiCompiler') (just as an example, dont know that if the name is correct).

I removed that check temporarily since the 2 other checks for client and server names in an array of 2 objects seemed sufficient.

I also had to change and add 
```
    multiCompiler.watch({
        aggregateTimeout: 300,
    }, (err, multiStats) => {
        debug('watch callback fired');
    });
```

to trigger the compiler.

**Comment by pho3nixf1re** [comment]
The instance check is unfortunate but I think that should go in another PR as it is unrelated to Koa support. Where did you add the watch? Is that a Koa specific thing or something with the babel-node setup you are running with?

**Comment by roastlechon** [comment]
i added the watch within the middleware to trigger the compiler.

Example:

```
multiCompiler.watch({
        aggregateTimeout: 300,
    }, (err, multiStats) => {
        debug('watch callback fired');
    });

    multiCompiler.plugin('done', multiStats => {
        debug('done compiling');
        error = false;
        const clientStats = findStats(multiStats, 'client');
        const serverStats = findStats(multiStats, 'server');
        // Server compilation errors need to be propagated to the client.
        if (serverStats.compilation.errors.length) {
            error = serverStats.compilation.errors[0];
            return;
        }
        const filename = getFilename(serverStats, outputPath, options.chunkName);
        const buffer = outputFileSystem.readFileSync(filename);
        const serverRendererOptions = Object.assign({
            clientStats: clientStats.toJson(),
            serverStats: serverStats.toJson()
        }, options.serverRendererOptions);
        try {
            serverRenderer = getServerRenderer(filename, buffer, serverRendererOptions);
        } catch (ex) {
            debug(ex);
            error = ex;
        }
    });
```

**Comment by roastlechon** [comment]
My requirements were support for koa@latest, webpack@latest, and being able to use it in conjunction with koa-webpack (webpack-dev-middleware and webpack-hot-middleware)

**Comment by pho3nixf1re** [comment]
Custom server handlers are now supported with the default set to the Express style and a Koa version provided for convenience. A release is pending but the change is on master. See #40

**Comment by richardscarrott** [comment]
Released in v0.2.0
